WordPress/gutenberg
# [TypeScript] Preserve the generic signature of getEntityRecord and getEntityRecords through currying

Most of the code proposed here once was a part of trunk. It got reverted in a large cleanup PR to reduce the complexity of the types, but unfortunately there is no way around the complex currying method proposed here.

## What problem does this PR solve?

`select( coreStore ).getEntityRecord()` has an incorrect TypeScript signature:

```ts
(kind: string, name: string, key: EntityRecordKey, query?: GetRecordsHttpQuery | undefined) => EntityRecord
```

It's supposed to be the same as the signature of `getEntityRecord` declared in selectors.ts, minus the state argument:

```ts
<
	EntityRecord extends
		| ET.EntityRecord<any>
		| Partial<ET.EntityRecord<any>>
>(
	state: State,
	kind: string,
	name: string,
	key: EntityRecordKey,
	query?: GetRecordsHttpQuery
) => EntityRecord | undefined;
```

This is because [TypeScript doesn't correctly curry generic functions](https://github.com/WordPress/gutenberg/pull/41578):

```ts
type CurriedState = F extends ( state: any, ...args: infer P ) => infer R
    ? ( ...args: P ) => R
    : F;

type Selector = <K extends string | number>(
    state: any,
    kind: K,
    key: K extends string ? 'string value' : false
) => K;

type BadlyInferredSignature = CurriedState< Selector >
// BadlyInferredSignature evaluates to:
// (kind: string number, key: false | "string value") => string number
```

## How does this PR propose to solve it?

This PR explicitly specifies both the curried and the uncurried signature. [This is the only working solution I found](https://github.com/WordPress/gutenberg/pull/41578).

Both signatures are shipped via the `GetEntityRecord` interface:

```ts
export interface GetEntityRecord {
	// Callable signature:
	< EntityRecord >( state: State, /* args */ ): EntityRecord | undefined;

	// Artificial property to couple with the callable signature:
	CurriedSignature: < EntityRecord >( /* args */ ): EntityRecord | undefined;
}
```

Then, the currying code in `@wordpress/data` gives `CurriedSignature` precedence over automatically currying the selector.

This commit also includes a docgen update to add support for typecasted selectors, otherwise CI shows a fatal error.

## Testing Instructions
Create the following test file and confirm it behaves in line with the screenshots below:

```ts
// Create a temporary file: packages/core-data/src/test.ts
import { useSelect } from '@wordpress/data';
import { store } from './index';
import type { Attachment } from './entity-types';

export function ReactComponent() {
    const attachment = useSelect(store, []).getEntityRecord<Attachment<'edit'>>('', '', '');
    return attachment?.title.rendered;
}
```

## Screenshots or screencast 

**Before:**
<img width="1041" alt="CleanShot 2022-09-26 at 18 54 54@2x" src="https://user-images.githubusercontent.com/205419/192235242-a1d72c97-fc20-47fe-9407-a54027de0ed9.png">

**After:**
<img width="901" alt="CleanShot 2022-09-26 at 18 46 51@2x" src="https://user-images.githubusercontent.com/205419/192234985-b0889850-9f3a-4e6f-9df9-aca368ea0edf.png">

cc @dmsnell @gziolo @noisysocks
